driver-core/src/main/com/mongodb/internal/connection/IamAuthenticator.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import com.mongodb.AuthenticationMechanism;
+import com.mongodb.MongoCredential;
+import com.mongodb.ServerAddress;
+import org.bson.BsonBinary;
+import org.bson.BsonBinaryWriter;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonString;
+import org.bson.ByteBuf;
+import org.bson.RawBsonDocument;
+import org.bson.codecs.BsonDocumentCodec;
+import org.bson.codecs.EncoderContext;
+import org.bson.io.BasicOutputBuffer;
+
+import javax.security.sasl.SaslClient;
+import javax.security.sasl.SaslException;
+import java.io.UnsupportedEncodingException;
+import java.security.SecureRandom;
+import java.util.Arrays;
+import java.util.Random;
+
+import static java.lang.String.format;
+
+public class IamAuthenticator extends SaslAuthenticator {
+    private static final int RANDOM_LENGTH = 32;
+    private final RandomStringGenerator randomStringGenerator;
+
+    public IamAuthenticator(final MongoCredentialWithCache credential) {
+        this(credential, new DefaultRandomStringGenerator());
+    }
+
+    public IamAuthenticator(final MongoCredentialWithCache credential, final RandomStringGenerator randomStringGenerator) {
+        super(credential);
+        this.randomStringGenerator = randomStringGenerator;
+    }
+
+    @Override
+    public String getMechanismName() {
+        AuthenticationMechanism authMechanism = getMongoCredential().getAuthenticationMechanism();
+        if (authMechanism == null) {
+            throw new IllegalArgumentException("Authentication mechanism cannot be null");
+        }
+        return authMechanism.getMechanismName();
+    }
+
+    @Override
+    protected SaslClient createSaslClient(final ServerAddress serverAddress) {
+        return new IamSaslClient(getMongoCredential(), randomStringGenerator);
+    }
+
+    public interface RandomStringGenerator {
+        String generate(int length);
+    }
+
+    private static class DefaultRandomStringGenerator implements RandomStringGenerator {
```

Comment:
@jyemin This was taken from ScramShaAuthenticator.java. Is there another class I could move this code to so I'm not duplicating it here? Same question for AuthorizationHeader::decodeUTF8

Reply:
From @markbenvenuto's comment below, seems like we won't want to share this after all.

Reply:
I wouldn't worry about decodeUTF8 since it's essentially a one-liner

Reply:
Done